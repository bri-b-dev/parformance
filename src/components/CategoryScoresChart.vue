<template>
  <section class="card" :aria-label="title || 'Kategorien'">
    <header style="display:flex; align-items:center; justify-content:space-between; gap:8px; margin-bottom:6px;">
      <h3 v-if="title" style="margin:0; font-weight:700;">{{ title }}</h3>
      <small class="chip" aria-label="Anzahl Kategorien">{{ categories.length }} Kategorien</small>
    </header>

    <!-- SR-only textual summary for accessibility -->
    <output class="sr-only" aria-live="polite">
      {{ srSummary }}
    </output>

    <!-- Bars (mobile) -->
    <div v-if="actualMode === 'bar'" class="row" style="flex-direction:column; gap:8px;">
      <div v-for="row in rows" :key="row.category" class="w-full">
        <div style="display:flex; align-items:center; justify-content:space-between; font-size:12px; margin-bottom:4px;">
          <span>{{ row.category }}</span>
          <span>{{ row.value }}%</span>
        </div>
        <div class="w-full" style="height:10px; background: var(--border); border-radius: 999px; overflow:hidden;">
          <div :data-testid="`bar-${row.category}`"
               :class="['h-full']"
               :style="barStyle(row)"></div>
        </div>
      </div>
    </div>

    <!-- Radar (desktop) -->
    <div v-else class="w-full" style="display:flex; align-items:center; justify-content:center;">
      <svg :width="svgSize" :height="svgSize" :viewBox="`0 0 ${svgSize} ${svgSize}`" role="img" aria-label="Radar Chart">
        <!-- grid circles -->
        <g :transform="`translate(${center}, ${center})`" stroke="var(--border)" fill="none">
          <circle v-for="r in [0.25, 0.5, 0.75, 1]" :key="r" :r="r * radius" />
        </g>
        <!-- axes -->
        <g :transform="`translate(${center}, ${center})`" stroke="var(--border)">
          <line v-for="(cat, i) in categories" :key="cat" :x1="0" :y1="0" :x2="axisPoint(i).x" :y2="axisPoint(i).y" />
        </g>
        <!-- polygon for values -->
        <g :transform="`translate(${center}, ${center})`">
          <polygon :points="polygonPoints" :fill="polyFill" :stroke="polyStroke" stroke-width="2" />
          <!-- weakest point marker -->
          <circle v-if="weakestIndex >= 0" :cx="valuePoint(weakestIndex).x" :cy="valuePoint(weakestIndex).y" r="4" fill="#ef4444" />
        </g>
        <!-- labels around -->
        <g font-size="11" fill="currentColor">
          <template v-for="(cat, i) in categories" :key="`lbl-${cat}`">
            <text :x="labelPoint(i).x" :y="labelPoint(i).y" text-anchor="middle" dominant-baseline="central">
              {{ cat }} ({{ values[i] }}%)
            </text>
          </template>
        </g>
      </svg>
    </div>
  </section>
</template>

<script setup lang="ts">
import { computed, onMounted, ref } from 'vue'

const props = defineProps<{ 
  scores: Record<string, number>,
  title?: string,
  /** 'auto' picks based on viewport; tests/stories can force 'bar' or 'radar' */
  mode?: 'auto' | 'bar' | 'radar',
}>()

const internalMode = ref<'bar'|'radar'>('bar')

onMounted(() => {
  if (props.mode === 'auto' || props.mode == null) {
    try {
      const w = typeof window !== 'undefined' ? window.innerWidth : 0
      internalMode.value = w >= 800 ? 'radar' : 'bar'
    } catch { internalMode.value = 'bar' }
  }
})

const actualMode = computed(() => {
  if (props.mode && props.mode !== 'auto') return props.mode
  return internalMode.value
})

const categories = computed(() => Object.keys(props.scores || {}))
const values = computed(() => categories.value.map(c => clampPct(props.scores[c])))

const rows = computed(() => {
  const out = categories.value.map((c, i) => ({ category: c, value: values.value[i] }))
  // highlight weakest by ordering? keep original; styling uses index
  return out
})

function clampPct(v: any): number {
  const n = Math.round(Number(v) || 0)
  return Math.max(0, Math.min(100, n))
}

const weakestIndex = computed(() => {
  let idx = -1; let min = Infinity
  values.value.forEach((v, i) => { if (v < min) { min = v; idx = i } })
  return idx
})

function barStyle(row: {category: string, value: number}) {
  const isWeak = categories.value.indexOf(row.category) === weakestIndex.value
  const color = isWeak ? '#ef4444' /* red-500 */ : '#2F7A52' /* primary */
  return {
    width: `${row.value}%`,
    background: color,
    borderRadius: '999px',
  }
}

// Radar geometry
const svgSize = 320
const center = svgSize / 2
const radius = Math.min(svgSize, svgSize) * 0.35

function angleFor(i: number) {
  const n = Math.max(1, categories.value.length)
  return -Math.PI / 2 + (i * 2 * Math.PI) / n
}

function axisPoint(i: number) {
  const a = angleFor(i)
  return { x: center + Math.cos(a) * radius, y: center + Math.sin(a) * radius }
}

function valuePoint(i: number) {
  const a = angleFor(i)
  const r = (values.value[i] / 100) * radius
  return { x: Math.cos(a) * r, y: Math.sin(a) * r }
}

function labelPoint(i: number) {
  const a = angleFor(i)
  const r = radius + 18
  return { x: center + Math.cos(a) * r, y: center + Math.sin(a) * r }
}

const polygonPoints = computed(() => {
  return categories.value.map((_c, i) => {
    const p = valuePoint(i)
    return `${center + p.x},${center + p.y}`
  }).join(' ')
})

const polyFill = computed(() => 'rgba(47,122,82,0.15)')
const polyStroke = computed(() => '#2F7A52')

const srSummary = computed(() => {
  if (categories.value.length === 0) return 'Keine Kategorien vorhanden.'
  const parts = categories.value.map((c, i) => `${c}: ${values.value[i]} Prozent`)
  const weak = weakestIndex.value >= 0 ? categories.value[weakestIndex.value] : null
  return `Fähigkeiten je Kategorie. ${parts.join(', ')}. Schwächste Kategorie: ${weak ?? '—'}.`
})
</script>
