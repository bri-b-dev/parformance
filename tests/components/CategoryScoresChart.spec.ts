import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import CategoryScoresChart from '../../src/components/CategoryScoresChart.vue'

const scores = {
  'Putting Green': 100,
  'Chipping Green': 33,
  'Driving Range': 67,
}

describe('CategoryScoresChart', () => {
  it('renders bar chart with labels, values, and highlights weakest', async () => {
    const wrapper = mount(CategoryScoresChart, { props: { scores, title: 'Scores', mode: 'bar' } })

    // Labels and values
    expect(wrapper.text()).toContain('Putting Green')
    expect(wrapper.text()).toContain('100%')
    expect(wrapper.text()).toContain('Chipping Green')
    expect(wrapper.text()).toContain('33%')

    // Weakest is Chipping Green (33). The bar should have red background (#ef4444)
    const bar = wrapper.get('[data-testid="bar-Chipping Green"]')
    const style = (bar.element as HTMLElement).getAttribute('style') || ''
    const normalized = style.replace(/\s+/g, ' ')
    expect(
      normalized.includes('background: rgb(239, 68, 68)') ||
      normalized.includes('background: #ef4444') ||
      normalized.includes('background:#ef4444')
    ).toBe(true)

    // SR summary present
    const sr = wrapper.find('p.sr-only')
    expect(sr.exists()).toBe(true)
    expect(sr.text()).toMatch(/SchwÃ¤chste Kategorie: Chipping Green/)

    // Snapshot of bar variant
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders radar chart with polygon and marker on weakest point', async () => {
    const wrapper = mount(CategoryScoresChart, { props: { scores, title: 'Scores', mode: 'radar' } })

    // Has an SVG and a polygon
    expect(wrapper.find('svg').exists()).toBe(true)
    expect(wrapper.find('polygon').exists()).toBe(true)

    // Weakest point marker exists (red circle)
    const marker = wrapper.find('circle[fill="#ef4444"]')
    expect(marker.exists()).toBe(true)

    // Snapshot of radar variant
    expect(wrapper.html()).toMatchSnapshot()
  })
})
