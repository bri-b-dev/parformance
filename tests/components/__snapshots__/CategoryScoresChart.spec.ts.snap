// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryScoresChart > renders bar chart with labels, values, and highlights weakest 1`] = `
"<section class="card" aria-label="Scores">
  <header style="display: flex; align-items: center; justify-content: space-between; gap: 8px; margin-bottom: 6px;">
    <h3 style="margin: 0px; font-weight: 700;">Scores</h3><small class="chip" aria-label="Anzahl Kategorien">3 Kategorien</small>
  </header><!-- SR-only textual summary for accessibility -->
  <p class="sr-only" role="status" aria-live="polite">F채higkeiten je Kategorie. Putting Green: 100 Prozent, Chipping Green: 33 Prozent, Driving Range: 67 Prozent. Schw채chste Kategorie: Chipping Green.</p><!-- Bars (mobile) -->
  <div class="row" style="flex-direction: column; gap: 8px;">
    <div class="w-full">
      <div style="display: flex; align-items: center; justify-content: space-between; font-size: 12px; margin-bottom: 4px;"><span>Putting Green</span><span>100%</span></div>
      <div class="w-full" style="height: 10px; background: var(--border); border-radius: 999px; overflow: hidden;">
        <div data-testid="bar-Putting Green" class="h-full" style="width: 100%; background: rgb(47, 122, 82); border-radius: 999px;"></div>
      </div>
    </div>
    <div class="w-full">
      <div style="display: flex; align-items: center; justify-content: space-between; font-size: 12px; margin-bottom: 4px;"><span>Chipping Green</span><span>33%</span></div>
      <div class="w-full" style="height: 10px; background: var(--border); border-radius: 999px; overflow: hidden;">
        <div data-testid="bar-Chipping Green" class="h-full" style="width: 33%; background: rgb(239, 68, 68); border-radius: 999px;"></div>
      </div>
    </div>
    <div class="w-full">
      <div style="display: flex; align-items: center; justify-content: space-between; font-size: 12px; margin-bottom: 4px;"><span>Driving Range</span><span>67%</span></div>
      <div class="w-full" style="height: 10px; background: var(--border); border-radius: 999px; overflow: hidden;">
        <div data-testid="bar-Driving Range" class="h-full" style="width: 67%; background: rgb(47, 122, 82); border-radius: 999px;"></div>
      </div>
    </div>
  </div>
</section>"
`;

exports[`CategoryScoresChart > renders radar chart with polygon and marker on weakest point 1`] = `
"<section class="card" aria-label="Scores">
  <header style="display: flex; align-items: center; justify-content: space-between; gap: 8px; margin-bottom: 6px;">
    <h3 style="margin: 0px; font-weight: 700;">Scores</h3><small class="chip" aria-label="Anzahl Kategorien">3 Kategorien</small>
  </header><!-- SR-only textual summary for accessibility -->
  <p class="sr-only" role="status" aria-live="polite">F채higkeiten je Kategorie. Putting Green: 100 Prozent, Chipping Green: 33 Prozent, Driving Range: 67 Prozent. Schw채chste Kategorie: Chipping Green.</p><!-- Bars (mobile) -->
  <!-- Radar (desktop) -->
  <div class="w-full" style="display: flex; align-items: center; justify-content: center;"><svg width="320" height="320" viewBox="0 0 320 320" role="img" aria-label="Radar Chart">
      <!-- grid circles -->
      <g transform="translate(160, 160)" stroke="var(--border)" fill="none">
        <circle r="28"></circle>
        <circle r="56"></circle>
        <circle r="84"></circle>
        <circle r="112"></circle>
      </g><!-- axes -->
      <g transform="translate(160, 160)" stroke="var(--border)">
        <line x1="0" y1="0" x2="160" y2="48"></line>
        <line x1="0" y1="0" x2="256.99484522385717" y2="215.99999999999997"></line>
        <line x1="0" y1="0" x2="63.00515477614289" y2="216.00000000000003"></line>
      </g><!-- polygon for values -->
      <g transform="translate(160, 160)">
        <polygon points="160,48 192.00829892387287,178.48 95.01345370001573,197.52000000000004" fill="rgba(47,122,82,0.15)" stroke="#2F7A52" stroke-width="2"></polygon><!-- weakest point marker -->
        <circle cx="32.00829892387286" cy="18.479999999999993" r="4" fill="#ef4444"></circle>
      </g><!-- labels around -->
      <g font-size="11" fill="currentColor"><text x="160" y="30" text-anchor="middle" dominant-baseline="central">Putting Green (100%) </text><text x="272.58330249197707" y="224.99999999999997" text-anchor="middle" dominant-baseline="central">Chipping Green (33%) </text><text x="47.41669750802299" y="225.00000000000006" text-anchor="middle" dominant-baseline="central">Driving Range (67%) </text></g>
    </svg></div>
</section>"
`;
